File: google-datacatalog-mysql-connector/src/google/datacatalog_connectors/mysql_/lineage_synchronizer/scrape/parse/extractLineage/src/main/java/com/sql/transform/Commons.java
package com.sql.transform;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Commons {
    private static boolean balancedString(String text) {
        int count = (int) text.chars().filter(num -> num == '\"').count();
        if(count % 2 == 1) {
            return false;
        }

        count = (int) text.chars().filter(num -> num == '\'').count();
        if(count % 2 == 1) {
            return false;
        }

        Stack<String> stack = new Stack<>();

        for(int i=0; i < text.length();i++) {
            String ch = text.charAt(i) + "";
            if(ch.equals("(") || ch.equals('[')) {
                stack.push(ch);
            } else if( ch.equals(")")) {
                if(!stack.empty() && stack.peek().equals("(")) {
                    stack.pop();
                } else {
                    return false;
                }
            } else if(ch.equals("]")) {
                if(!stack.empty() && stack.peek().equals("[")) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }

        if(!stack.empty()) {
            return false;
        }

        return true;
    }

        public static int findSqlKeyword(String text, String target) {
        Matcher m;
        m = Pattern.compile(target, Pattern.CASE_INSENSITIVE).matcher(text);
        List<Integer> posList = new ArrayList<Integer>();
        while (m.find())
        {
            posList.add(m.start());
        }

        for(int pos: posList) {
            String beforeFrom = text.substring(0,pos);
            if(balancedString(beforeFrom)) {
                return pos;
            }
        }

        return -1;
    }
}

File: google-datacatalog-mysql-lineage-extractor/src/google/datacatalog_connectors/mysql_/parse/transform_equ/commons.py
import re

class Commons:
    @staticmethod
    def _balanced_text(text: str) -> str:
        c1 = text.count('\'')
        c2 = text.count('\"')
        
        if c1%2 != 0 or c2%2 != 0:
            return False
        
        stack = []
        
        for ch in text:
            if ch == '(' or ch == '[':
                stack.append(ch) 
            elif ch == ')':
                if len(stack) > 0 and stack[-1] == '(':
                    stack.pop()
                else:
                    return False
                
            elif ch == ']':
                if len(stack) > 0 and stack[-1] == '[':
                    stack.pop()
                else:
                    return False
            
        if len(stack) != 0:
            return False

        return True
    
    @staticmethod
    def findSqlKeyword(text: str,target: str):
        target = '\\w*(?<![a-z0-9_])' + target + '[\\s\\(]';
        regex = re.compile(target, re.IGNORECASE)
        
        for m in regex.finditer(text):
            matchPos = m.start()
            if(Commons._balanced_text(text[0:matchPos])):
                return matchPos
            
        return None